### PR TITLE
Refactor monitor module for handling netlink events

### PR DESCRIPTION
**Pull request by pperiyasamy**
This PR refactors handling of hostapd events into go module which also helps handling netlink events for the interfaces  serving EAP authentication. It also provisions hostapd for TLS configuration, sample certificates and wpa_supplicant client script samples which can be useful for testing.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -244,7 +244,7 @@ catalog-push: ## Push a catalog image.
 
 .PHONY: authenticator-build
 authenticator-build: ## Build the authenticator image.
-	docker build -f ./images/eapol-authenticator/Dockerfile -t $(AUTHENTICATOR_IMG) ./images/eapol-authenticator/
```
Comment (lack) [inline Makefile:247]: I don't see a removal of this older Dockerfile in this PR - Should we do that too?

File: internal/logging/logging.go
Comment (lack) [inline internal/logging/logging.go:1]: This looks like something from some other library - Is it better to reference it as opposed to copying it?

File: internal/sriov/sriov.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	return vfNames, nil
+}
+
+func UnprotectPorts(logger log.Logger, ifName string, protocol string, ports []int) error {
+	if _, err := exec.LookPath("tc"); err != nil {
+		return err
+	}
+	for _, port := range ports {
+		cmd := exec.Command("bash", "-c", fmt.Sprintf("tc filter add dev %s ingress pref 9999 protocol ip u32 match %s dst %s 0xffff action ok index 99", ifName, protocol, strconv.Itoa(port)))
```
Comment (lack) [inline internal/sriov/sriov.go]: Do we need to invoke bash?  I think we could just run the tc command directly, couldn't we?

File: monitor/monitor.go
Comment (lack) [inline monitor/monitor.go:1]: May I suggest calling this `hostapMonitor` instead?  And maybe take it out of the `main` package for potential future reuse, like you did with the `sriov` package?  I could see this being in its own library some day...

File: pkg/configgen/configgen.go [hunk 3]
```diff
@@ -130,6 +134,8 @@ func (g *ConfigGenerator) Daemonset() *appsv1.DaemonSet {
 			},
 		})
 	}
+	projectedConfigVolumes = g.injectCertVolumes(projectedConfigVolumes)
+	projectedConfigVolumes = g.injectRadiusClientVolume(projectedConfigVolumes)
```
Comment (lack) [inline pkg/configgen/configgen.go]: I found this a bit confusing - Can we rename these functions to `append*Volume[s]`

File: pkg/configgen/configgen.go [hunk 4]
```diff
@@ -209,6 +215,17 @@ func (g *ConfigGenerator) Daemonset() *appsv1.DaemonSet {
...
+							[]corev1.EnvVar{{
+								Name:  "IFACES",
+								Value: ifaces,
+							}, {
+								Name:  "UNPROTECTED_TCP_PORTS",
+								Value: unprotectedTcpList,
+							}, {
+								Name:  "UNPROTECTED_UDP_PORTS",
+								Value: unprotectedUdpList,
+							}}),
```
Comment (lack) [inline pkg/configgen/configgen.go:228]: Yay!  Exactly one container!!

File: monitor/monitor.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			if newOpState == m.operState {
+				level.Info(m.logger).Log("interface", "event", m.ifName, "op state not changed", m.operState)
+				continue
+			}
+			m.operState = newOpState
+			level.Info(m.logger).Log("interface", "event", m.ifName, "op state changed", m.operState)
+			m.mutex.Lock()
+			if newOpState == netlink.OperUp {
+				for mac := range m.authenticatedMacs {
+					err := sriov.AllowTrafficFromMac(m.ifName, mac)
```
Comment (lack) [inline monitor/monitor.go]: Does this handle the PF as well, or just the VFs?

File: internal/sriov/sriov.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+func IsSriovPF(ifName string) bool {
+	ifPfDir := filepath.Join(sysClassNet, ifName, "device", "physfn")
+	if _, err := os.Stat(ifPfDir); err != nil {
+		return false
+	}
+	return true
+}
+
+func GetSriovVFs(ifName string) ([]string, error) {
```
Comment (lack) [inline internal/sriov/sriov.go]: Is this called anywhere?  It looks like we're missing logic to actual interact with the VFs.

File: internal/sriov/sriov.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+		err := cmd.Run()
+		if err != nil {
+			level.Error(logger).Log("op", "tc filter add", "ifName", ifName, "protocol", protocol, "port", port, "error", err)
+		}
+		// TODO: Need to figure out what to do for ipv6, as tc doesn't support it!
+	}
+	return nil
+}
+
+func InitInterfaceForEAPTraffic(logger log.Logger, ifName string, unprotectTcpPorts, unprotectUdpPorts []int) error {
```
Comment (lack) [inline internal/sriov/sriov.go]: Is this called anywhere?  I couldn't find any code that actually sets up the interfaces.

File: internal/sriov/sriov.go
Comment (lack) [inline internal/sriov/sriov.go:1]: Maybe a better name for this class would be "Traffic Control"? It's not really specific to SRIOV, except that we need to deal with SRIOV devices?

**Review by lack — CHANGES_REQUESTED** [review]
Other than some trivial requests for renaming files or functions, I think the 2 main issues with this PR are:
- It does not set up the interface to block all (except EAPOL and "allowed ports") when the monitor starts up
- It does not allow traffic on the VFs or deny traffic on the VFs when hostapd authentication/deauthentication, it seems to only operate on the PFs.

File: api/v1/authenticator_types.go [hunk 8]
```diff
@@ -75,6 +75,29 @@ type Local struct {
...
+	// +optional
+	PrivateKeyPassphrase string `json:"privateKeyPassphrase,omitempty"`
+	// RadiusClientSecret secret reference containing client information for local radius server.
+	// If the key is not specified, it is assumed to be "hostapd.radius_clients"
+	// +optional
+	RadiusClientSecret *SecretKeyRef `json:"radiusClientFileSecret,omitempty"`
+	// AuthPort UDP listening port Local Radius authentication server.
+	// +kubebuilder:default=1812
+	// +optional
+	AuthPort int `json:"authPort"`
```
Comment (lack) [inline api/v1/authenticator_types.go:100]: Are these mutually exclusive with `UserFileSecret`, or can/should they be used in conjunction with it?

File: Makefile (hunk 1 above)
Comment (pperiyasamy) [reply Makefile:247]: yes, removed older Dockerfile and hostapd-cli.sh script which is not needed.

File: api/v1/authenticator_types.go (hunk 8 above)
Comment (pperiyasamy) [reply api/v1/authenticator_types.go:100]: yes, both are mutually exclusive.  but all comes under local EAP auth configuration for hostapd.

File: internal/sriov/sriov.go (hunk 2 above)
Comment (pperiyasamy) [reply internal/sriov/sriov.go]: yes, without using `bash`, tc command throws path not found error which its run from exec.

File: monitor/monitor.go
Comment (pperiyasamy) [reply monitor/monitor.go:1]: sure, done.

File: pkg/configgen/configgen.go (hunk 3 above)
Comment (pperiyasamy) [reply pkg/configgen/configgen.go]: changed it.

File: monitor/monitor.go (hunk 5 above)
Comment (pperiyasamy) [reply monitor/monitor.go]: oh! yes, now changes are made to handle VFs as well.

File: internal/sriov/sriov.go (hunk 6 above)
Comment (pperiyasamy) [reply internal/sriov/sriov.go]: now it's used for allow/deny traffic cases.

File: internal/sriov/sriov.go (hunk 7 above)
Comment (pperiyasamy) [reply internal/sriov/sriov.go]: This is just a placeholder method which might get used  when we move iface-init and hostapd containers into same monitor container.

File: internal/logging/logging.go
Comment (pperiyasamy) [reply internal/logging/logging.go:1]: yes, this is reused from metallb internal package. added some description about this in the logging module now.

File: internal/sriov/sriov.go
Comment (pperiyasamy) [reply internal/sriov/sriov.go:1]: sure, changed it.

**Comment by pperiyasamy** [comment]
> * It does not set up the interface to block all (except EAPOL and "allowed ports") when the monitor starts up

This is already being done as part of iface-init container, so not redoing it from monitor.

> * It does not allow traffic on the VFs or deny traffic on the VFs when hostapd authentication/deauthentication, it seems to only operate on the PFs.

Now changes are made to handle PF and its corresponding VFs as well.

File: internal/trafficcontrol/tc.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+func AllowTrafficFromMac(ifName string, macAddress string) error {
+	if _, err := exec.LookPath("tc"); err != nil {
+		return err
+	}
+	interfaces, err := GetAssociatedInterfaces(ifName)
+	if err != nil {
+		return err
+	}
+	for _, iface := range interfaces {
+		cmd := exec.Command("bash", "-c", fmt.Sprintf("tc filter replace dev %s ingress pref 9000 protocol all flower src_mac %s action ok", iface, macAddress))
```
Comment (lack) [inline internal/trafficcontrol/tc.go:47]: I think we need to eliminate bash from all of these exec commands.  If you use `/usr/sbin/tc` instead of `tc`, does that work?

That said, I'm okay with that kind of cleanup being in a separate PR; no need to block this one on that change provided we plan to clean this up in the future.

File: pkg/hostap/monitor.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+		case linkUpdateEvent := <-m.ifEventCh:
+			newOpState := linkUpdateEvent.Link.Attrs().OperState
+			if newOpState == m.operState {
+				level.Info(m.Logger).Log("interface", "event", m.IfName, "op state not changed", m.operState)
+				continue
+			}
+			m.operState = newOpState
+			level.Info(m.Logger).Log("interface", "event", m.IfName, "op state changed", m.operState)
+			m.mutex.Lock()
+			if newOpState == netlink.OperUp {
```
Comment (lack) [inline pkg/hostap/monitor.go]: I'm not sure if this logic is correct...

If a MAC is previously authenticated and the link goes down, we want to signal hostAPD that it's deauthenticated - If hostapd sends us a deauthenticated message, we can use handling of that message to block traffic.  If it doesn't, we'll have to additionally deny the traffic ourselves.  And we should flush `m.authenticatedMacs`.  When a link down happens, any previous MACs that had authenticated are no longer authenticated.

So that means that the bigger issue is what happens when the link comes back up... We do not want to immediately allow traffic from it again - We must wait for hostapd to do the actual 802.1x authentication and signal us that it's properly authenticated.  So we shouldn't ever call AlowTrafficFromMac in this handler, but only in the message handler for hostapd events.

File: pkg/hostap/monitor.go [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+			if err != nil {
+				level.Error(m.Logger).Log("hostapd-event", eventStr, "interface", m.IfName, "error", err)
+			}
+		}
+	}
+}
+
+func (m *InterfaceMonitor) handleHostapdEvent(eventStr string) error {
+	eventStrSlice := strings.Split(eventStr, " ")
+	if len(eventStrSlice) < 2 {
```
Comment (lack) [inline pkg/hostap/monitor.go:141]: Probably we should log unexpectedly short messages?  Or do we already log all messages?

File: pkg/hostap/monitor.go [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+func (m *InterfaceMonitor) handleAutheticatedMac(macAddress string) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	m.authenticatedMacs[macAddress] = nil
+	err := trafficcontrol.AllowTrafficFromMac(m.IfName, macAddress)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (lack) [inline pkg/hostap/monitor.go]:
```suggestion
	return trafficcontrol.AllowTrafficFromMac(m.IfName, macAddress)
```
Isn't this the same, as long as appropriate logging already occurs in AllowTrafficFromMac?

File: internal/trafficcontrol/tc.go [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+		cmd := exec.Command("bash", "-c", fmt.Sprintf("tc filter replace dev %s ingress pref 9000 protocol all flower src_mac %s action drop", iface, macAddress))
+		err := cmd.Run()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func InitInterfaceForEAPTraffic(logger log.Logger, ifName string, unprotectTcpPorts, unprotectUdpPorts []int) error {
```
Comment (lack) [inline internal/trafficcontrol/tc.go:74]: Still not called anywhere, but I understand why.

I'd much rather have a second entrypoint or an `--init` flag in main.go to call this same code for the init container so all the 'tc' operations are in one file and not split across this file AND a shell script, but that can be in a separate PR.

File: internal/sriov/sriov.go (hunk 7 above)
Comment (lack) [reply internal/sriov/sriov.go]: I think we should make sure we add a task somewhere to do this.  It's an important cleanup.

File: internal/trafficcontrol/tc.go (hunk 9 above)
Comment (pperiyasamy) [reply internal/trafficcontrol/tc.go:47]: yes, i tried with using absolute path for tc command as well, that didn't work either. hence used "bash", "-c" option.

File: internal/trafficcontrol/tc.go (hunk 13 above)
Comment (pperiyasamy) [reply internal/trafficcontrol/tc.go:74]: sure, will do this with another PR and that'll completely remove using scripts.

File: pkg/hostap/monitor.go (hunk 11 above)
Comment (pperiyasamy) [reply pkg/hostap/monitor.go:141]: yes, now logging all the unhandled events.

File: pkg/hostap/monitor.go (hunk 10 above)
Comment (pperiyasamy) [reply pkg/hostap/monitor.go]: you are right. now sending deauth request for addresses upon interface down event, that would eventually deny traffic and flush the mac entry upon handling `AP-STA-DISCONNECTED` event. 
Also handling requests timeout with `handleRequestTimeouts` goroutine.

File: pkg/hostap/monitor.go (hunk 12 above)
Comment (pperiyasamy) [reply pkg/hostap/monitor.go]: yes, done.

**Review by lack — APPROVED** [review]
